### PR TITLE
[Do not merge] New config groups API endpoint

### PR DESCRIPTION
**Pull request by mapedraza**
Issue #752 
Overpasses #1375 

Kudos to: @KeshavSoni2511 

Pending:
>Just some minor improvements points:
>
> 1): I can't see tests implementing cross functionality. I mean, creating groups using services and then retrieving it through configGroup endpoint. It should be aded in both ways, I mean, creating as services and configGroup and retrieving them with the opposite.
> 2) In order to configure the keyword used by the new endpoint, it would be great if instead of having a magic word, it is moved to the constans file (lib/constants.js), replacing each time in the code appears 'configgroups'. It should be configured for the API route and for the term used in the json response.

Check before merging:

- [x] Modify just merged doc/getting-started.md to use the new API (Done here: https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/cdb791c77ee00f628d22753311042591ab259ec6)

### CODE REVIEW

**Comment by mapedraza** [comment]
Commit aa6457333463fdeb7878a07e94061843197a2a36 solves pending point 2:

>2) In order to configure the keyword used by the new endpoint, it would be great if instead of having a magic word, it is moved to the constans file (lib/constants.js), replacing each time in the code appears 'configgroups'. It should be configured for the API route and for the term used in the json response.

Tests use a constants strings (`http://localhost:4041/iot/configGroups`) instead of a parameter (`constants.CONFIGGROUP_TERM`) to be more legible

**Comment by mapedraza** [comment]
Commit https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/0b38688730b7f85780d25ab787de8d97551abef3  solves pending point 1):

> 1): I can't see tests implementing cross functionality. I mean, creating groups using services and then retrieving it through configGroup endpoint. It should be aded in both ways, I mean, creating as services and configGroup and retrieving them with the opposite.

File: package-lock.json
Comment (AlvaroVega) [inline package-lock.json:1]: This file should not be part of repository.

File: test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json
Comment (AlvaroVega) [inline test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json:1]: Bidirectional feature was not removed on previous PRs ?

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -3,4 +3,5 @@
 - Fix: service header to use uppercase in case of update and delete (#1528)
 - Fix: Allow to send to CB batch update for multimeasures for NGSI-LD (#1623)
 - Add: new JEXL transformations for including into an array keys that have a certain value: valuePicker and valuePickerMulti
+- Add /iot/configGroups API endpoints (as equivalent to /iot/services) (#752)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:6]:
```suggestion
- Add /iot/configGroups API endpoints (as equivalent to /iot/services) (#752)
- Deprecated: /iot/services API routes
```

File: doc/api.md [hunk 2]
```diff
@@ -1847,11 +1850,11 @@ _**Response headers**_
 
 Successful operations return `Content-Type` header with `application/json` value.
 
-#### Modify config group `PUT /iot/services`
+#### Modify config group `PUT /iot/configGroups` or `PUT /iot/services`
```
Comment (fgalan) [inline doc/api.md:1853]: I'd suggest to remove references to `/iot/services` in any file, as it is deprecated feature and, as stated in [doc](https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/deprecated.md)

> Documentation on deprecated features is removed from the repository documentation. Documentation is still available in the documentation set associated to older versions (in the repository release branches).

File: test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json
Comment (mapedraza) [reply test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json:1]: Added this PR with more leftovers: https://github.com/telefonicaid/iotagent-node-lib/pull/1644

Once it gets merged, this PR should be sync

File: package-lock.json
Comment (mapedraza) [reply package-lock.json:1]: Removed in https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/e2ea6c20fc114afe00c71335e52fa154541f57b4
Added to gitignore here: https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/d48cc4b6ed492961235bf4bee8b3070acccc7e9a

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (mapedraza) [reply CHANGES_NEXT_RELEASE:6]: Added here:https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/e1f4e47cccf25b7ace81e33f9eb4801bfd60335c

This commit also includes an entry in deprecated.md file

File: test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json
Comment (mapedraza) [reply test/unit/examples/groupProvisioningRequests/bidirectionalConfigGroup.json:1]: Done in: https://github.com/telefonicaid/iotagent-node-lib/pull/1637/commits/611b45fd9c4854b2d7fcb3daf80414f60329c2b4

File: doc/api.md (hunk 2 above)
Comment (mapedraza) [reply doc/api.md:1853]: Pending to choose the final term for `configGroups`

File: doc/deprecated.md [hunk 3]
```diff
@@ -26,6 +26,8 @@ A list of deprecated features and the version in which they were deprecated foll
 -   Bidirectinal pluging (finally removed in 3.4.0)
 -   appendMode configuration (`IOTA_APPEND_MODE` env var) (finally removed in 3.4.0)
 -   `config.stats` section, and push-mode statistics.
+-   Services API routes (`/iot/services`) in favor of the `/iot/configGroups`. Both are still supported, but the former
+    is deprecated.
```
Comment (fgalan) [inline doc/deprecated.md:30]:
```suggestion
-   Services API routes (`/iot/services`) in favor of the `/iot/configGroups`.
```

Looking the above elements, it is assumed that every deprecated feature is still supported, except the ones with the "(finally removed in x)" part.

**Review by AlvaroVega — APPROVED** [review]
LGTM

File: lib/fiware-iotagent-lib.js [hunk 4]
```diff
@@ -47,17 +47,20 @@ const context = {
...
+                deviceRemovalRequests: 0,
+                measureRequests: 0,
+                raiseAlarm: 0,
+                releaseAlarm: 0,
+                updateEntityRequestsOk: 0,
+                updateEntityRequestsError: 0
+            })
+        ],
+        callback
+    );
```
Comment (mrutid) [inline lib/fiware-iotagent-lib.js:63]: try to avoid "Linter changes" not corresponding to current PR.

**Comment by fgalan** [comment]
After @mrutid @mapedraza @fgalan discussion today, this PR is going to be re-worked.

**Comment by mapedraza** [comment]
overpassed by #1648

**Pull request closed without merging**